Mathlib/Data/Fin/Tuple/Reflection.lean

```diff
@@ -132,19 +132,85 @@ def sum [Add α] [Zero α] : ∀ {m} (_ : Fin m → α), α
   | 1, v => v 0
   | _ + 2, v => sum (fun i => v (Fin.castSucc i)) + v (Fin.last _)
 
+-- `to_additive` without `existing` fails, see
+-- https://leanprover.zulipchat.com/#narrow/channel/287929-mathlib4/topic/to_additive.20complains.20about.20equation.20lemmas/near/508910537
+/-- `Finset.univ.prod` with better defeq for `Fin`. -/
+@[to_additive existing]
+def prod [Mul α] [One α] : ∀ {m} (_ : Fin m → α), α
+  | 0, _ => 1
+  | 1, v => v 0
+  | _ + 2, v => prod (fun i => v (Fin.castSucc i)) * v (Fin.last _)
+
 /-- This can be used to prove
 ```lean
-example [AddCommMonoid α] (a : Fin 3 → α) : ∑ i, a i = a 0 + a 1 + a 2 :=
-  (sum_eq _).symm
+example [CommMonoid α] (a : Fin 3 → α) : ∏ i, a i = a 0 * a 1 * a 2 :=
+  (prod_eq _).symm
 ```
 -/
-@[simp]
-theorem sum_eq [AddCommMonoid α] : ∀ {m} (a : Fin m → α), sum a = ∑ i, a i
+@[to_additive (attr := simp)
+"This can be used to prove
+```lean
+example [AddCommMonoid α] (a : Fin 3 → α) : ∑ i, a i = a 0 + a 1 + a 2 :=
+  (sum_eq _).symm
+```"]
+theorem prod_eq [CommMonoid α] : ∀ {m} (a : Fin m → α), prod a = ∏ i, a i
   | 0, _ => rfl
-  | 1, a => (Fintype.sum_unique a).symm
-  | n + 2, a => by rw [Fin.sum_univ_castSucc, sum, sum_eq]
+  | 1, a => (Fintype.prod_unique a).symm
+  | n + 2, a => by rw [Fin.prod_univ_castSucc, prod, prod_eq]
+
+example [CommMonoid α] (a : Fin 3 → α) : ∏ i, a i = a 0 * a 1 * a 2 :=
+  (prod_eq _).symm
 
 example [AddCommMonoid α] (a : Fin 3 → α) : ∑ i, a i = a 0 + a 1 + a 2 :=
   (sum_eq _).symm
 
 end FinVec
+
+namespace Fin
+
+namespace ProdUnivMany
+
+open Lean Meta Qq
+
+/-- Implementation of the `prod_univ_many` simproc. -/
+def prodEquivManyImp {u : Level} {α : Q(Type u)} (inst : Q(CommMonoid $α)) (n : ℕ)
+    (f : Q(Fin $n → $α)) :
+    MetaM <| (val : Q($α)) × Q(∏ i, $f i = $val) := do
+  match n with
+  | 0 =>
+    return ⟨q((1 : $α)), q(Fin.prod_univ_zero $f)⟩
+  | m + 1 =>
+    let nezero : Q(NeZero ($m + 1)) := q(⟨Nat.succ_ne_zero _⟩)
+    let val ← makeRHS (m + 1) f nezero (m + 1)
+    let _ : $val =Q FinVec.prod $f := ⟨⟩
+    let pf := q(FinVec.prod_eq $f |>.symm)
+    return ⟨val, pf⟩
+where
+  /-- Creates the expression `f 0 * f 1 * ... * f (n - 1)`. -/
+  makeRHS (n : ℕ) (f : Q(Fin $n → $α)) (nezero : Q(NeZero $n))
+    (k : ℕ) : MetaM Q($α) := do
+  match k with
+  | 0 => failure
+  | 1 =>
+    pure q($f 0)
+  | m + 1 =>
+    let pre ← makeRHS n f nezero m
+    pure q($pre * $f (@OfNat.ofNat (Fin $n) $m _))
```

Comment:
I suspect this generates a bad term with an extra `ofNat`; in general, I'd be wary of interpolating naturals into Qq expressions, and would recommend using a `mkRawNatLit` to generate the inner numeral here.

`pp.natLit true` should make it clear whether the current code is problematic.